src/cluster_linearize.h

```diff
@@ -1336,6 +1347,38 @@ std::vector<ClusterIndex> MergeLinearizations(const DepGraph<SetType>& depgraph,
     return ret;
 }
 
+/** Make linearization topological, retaining its ordering where possible. */
+template<typename SetType>
+void FixLinearization(const DepGraph<SetType>& depgraph, Span<DepGraphIndex> linearization) noexcept
+{
+    // This algorithm can be summarized as moving every element in the linearization backwards
+    // until it is placed after all its ancestors.
+    SetType done;
+    const auto len = linearization.size();
+    // Iterate over the elements of linearization from back to front (i is distance from back).
+    for (DepGraphIndex i = 0; i < len; ++i) {
+        /** The element at that position. */
+        DepGraphIndex elem = linearization[len - 1 - i];
+        /** j represents how far from the back of the linearization elem should be placed. */
+        DepGraphIndex j = i;
+        // Figure out which elements need to be moved before elem.
+        SetType place_before = done & depgraph.Ancestors(elem);
+        // Find which position to place elem in (updating j), continuously moving the elements
+        // in between forward.
+        while (place_before.Any()) {
+            // j cannot be 0 here; if it was, then there was necessarily nothing earlier which
```

Comment:
```suggestion
            // j cannot be 0 or less; if it was, then there was necessarily nothing earlier which
```

Reply:
```
if it was, then there was necessarily nothing earlier which elem needs to be place before anymore, and place_before would be empty.
```
This comment seems jumbled and hard to understand.  Is it possible to word this better?